### ISSUE TITLE
[Concurrency] Menu query doesn't handle concurrency

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/Orchard#8040]
```
_contentManager.Query().ForType("Menu").Join<TitlePartRecord>().OrderBy(x => x.Title).List())
```

In MainMenuService.

**Comment by sebastienros** [comment]
For reference while trying to understand why this query is locking:
https://www.brentozar.com/archive/2018/04/how-entity-framework-and-nhibernate-can-cause-poison-resource_semaphore-waits/
http://sqlapproach.blogspot.com/2012/08/active-processes.html

**Comment by MatteoPiovanelli** [comment]
tl;dr: The issue we have seems to not be the one indicated in the links.

I am trying those tests.

I am using the 1.10.x branch, and a fresh tenant. I try to create contents of type Page. The definition of this ContentType has been changed, so it only contains: CommonPart, TitlePart, MenuPart. I did that because we know there is a concurrency issue with the MenuPart, and I did not want to have other stuff making things even more complex.

Concurrent creation of two items fails, as we knew it would. Here's the full error from the logs:

```
2018-05-10 09:06:01,454 [38] Orchard.ContentManagement.Drivers.Coordinators.ContentPartDriverCoordinator - DevNoLaser - ERROR GenericADOException thrown from IContentPartDriver by Orchard.Core.Navigation.Drivers.MenuPartDriver
NHibernate.Exceptions.GenericADOException: could not execute query
[ SELECT this_.Id as Id11_3_, this_.Number as Number11_3_, this_.Published as Published11_3_, this_.Latest as Latest11_3_, this_.Data as Data11_3_, this_.ContentItemRecord_id as ContentI6_11_3_, titlepartr2_.Id as Id136_0_, titlepartr2_.Title as Title136_0_, titlepartr2_.ContentItemRecord_id as ContentI3_136_0_, contentite1_.Id as Id7_1_, contentite1_.Data as Data7_1_, contentite1_.ContentType_id as ContentT3_7_1_, contenttyp6_.Id as Id13_2_, contenttyp6_.Name as Name13_2_ FROM Orchard_Framework_ContentItemVersionRecord this_ inner join Title_TitlePartRecord titlepartr2_ on this_.Id=titlepartr2_.Id inner join Orchard_Framework_ContentItemRecord contentite1_ on this_.ContentItemRecord_id=contentite1_.Id left outer join Orchard_Framework_ContentTypeRecord contenttyp6_ on contentite1_.ContentType_id=contenttyp6_.Id WHERE contentite1_.ContentType_id = @p0 and this_.Published = @p1 ORDER BY titlepartr2_.Title asc ]
  Name:cp0 - Value:1  Name:cp1 - Value:True
[SQL: SELECT this_.Id as Id11_3_, this_.Number as Number11_3_, this_.Published as Published11_3_, this_.Latest as Latest11_3_, this_.Data as Data11_3_, this_.ContentItemRecord_id as ContentI6_11_3_, titlepartr2_.Id as Id136_0_, titlepartr2_.Title as Title136_0_, titlepartr2_.ContentItemRecord_id as ContentI3_136_0_, contentite1_.Id as Id7_1_, contentite1_.Data as Data7_1_, contentite1_.ContentType_id as ContentT3_7_1_, contenttyp6_.Id as Id13_2_, contenttyp6_.Name as Name13_2_ FROM Orchard_Framework_ContentItemVersionRecord this_ inner join Title_TitlePartRecord titlepartr2_ on this_.Id=titlepartr2_.Id inner join Orchard_Framework_ContentItemRecord contentite1_ on this_.ContentItemRecord_id=contentite1_.Id left outer join Orchard_Framework_ContentTypeRecord contenttyp6_ on contentite1_.ContentType_id=contenttyp6_.Id WHERE contentite1_.ContentType_id = @p0 and this_.Published = @p1 ORDER BY titlepartr2_.Title asc] ---> System.Data.SqlClient.SqlException: Transaction (Process ID 68) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction.
   in System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   in System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   in System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   in System.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   in System.Data.SqlClient.SqlDataReader.TryReadInternal(Boolean setTimeout, Boolean& more)
   in System.Data.SqlClient.SqlDataReader.Read()
   in NHibernate.Driver.NHybridDataReader.Read()
   in NHibernate.Loader.Loader.DoQuery(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer)
   in NHibernate.Loader.Loader.DoQueryAndInitializeNonLazyCollections(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer)
   in NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer)
   --- Fine della traccia dello stack dell'eccezione interna ---
   in NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer)
   in NHibernate.Loader.Loader.ListIgnoreQueryCache(ISessionImplementor session, QueryParameters queryParameters)
   in NHibernate.Impl.SessionImpl.List(CriteriaImpl criteria, IList results)
   in NHibernate.Impl.CriteriaImpl.List(IList results)
   in NHibernate.Impl.CriteriaImpl.List[T]()
   in Orchard.ContentManagement.DefaultContentQuery.Slice(Int32 skip, Int32 count) in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard\ContentManagement\DefaultContentQuery.cs:riga 191
   in Orchard.ContentManagement.DefaultContentQuery.ContentQuery`1.Orchard.ContentManagement.IContentQuery<T>.List() in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard\ContentManagement\DefaultContentQuery.cs:riga 292
   in Orchard.Core.Navigation.Services.MainMenuService.GetMenus() in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard.Web\Core\Navigation\Services\MainMenuService.cs:riga 64
   in Orchard.Core.Navigation.Drivers.MenuPartDriver.Editor(MenuPart part, Object shapeHelper) in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard.Web\Core\Navigation\Drivers\MenuPartDriver.cs:riga 40
   in System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   in Orchard.ContentManagement.Drivers.ContentPartDriver`1.Orchard.ContentManagement.Drivers.IContentPartDriver.UpdateEditor(UpdateEditorContext context) in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard\ContentManagement\Drivers\ContentPartDriver.cs:riga 62
   in Orchard.ContentManagement.Drivers.Coordinators.ContentPartDriverCoordinator.<>c__DisplayClass11_0.<UpdateEditor>b__0(IContentPartDriver driver) in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard\ContentManagement\Drivers\Coordinators\ContentPartDriverCoordinator.cs:riga 63
   in Orchard.InvokeExtensions.Invoke[TEvents](IEnumerable`1 events, Action`1 dispatch, ILogger logger) in C:\Sviluppo\Orchard.Sources.Github\Orchard\src\Orchard\InvokeExtensions.cs:riga 17
```
Just the query (@p0 = 1 and @p1 = True):

```
SELECT 
	this_.Id as Id11_3_, 
	this_.Number as Number11_3_, 
	this_.Published as Published11_3_, 
	this_.Latest as Latest11_3_, 
	this_.Data as Data11_3_, 
	this_.ContentItemRecord_id as ContentI6_11_3_, 
	titlepartr2_.Id as Id136_0_, 
	titlepartr2_.Title as Title136_0_, 
	titlepartr2_.ContentItemRecord_id as ContentI3_136_0_, 
	contentite1_.Id as Id7_1_, 
	contentite1_.Data as Data7_1_, 
	contentite1_.ContentType_id as ContentT3_7_1_, 
	contenttyp6_.Id as Id13_2_, 
	contenttyp6_.Name as Name13_2_ 
FROM 
	Orchard_Framework_ContentItemVersionRecord this_ 
	inner join Title_TitlePartRecord titlepartr2_ on this_.Id=titlepartr2_.Id 
	inner join Orchard_Framework_ContentItemRecord contentite1_ on this_.ContentItemRecord_id=contentite1_.Id 
	left outer join Orchard_Framework_ContentTypeRecord contenttyp6_ on contentite1_.ContentType_id=contenttyp6_.Id 
WHERE 
	contentite1_.ContentType_id = @p0 
	and this_.Published = @p1 
ORDER BY titlepartr2_.Title asc
```

Following what was done at https://www.brentozar.com/archive/2018/04/how-entity-framework-and-nhibernate-can-cause-poison-resource_semaphore-waits/ , I copied that query in SQLQueryStress, and ran it at 50 iterations on 50 threads. 
The table for the WAIT_STATS of the results for that procedure had no RESOURCE_SEMAPHORE line:
![image](https://user-images.githubusercontent.com/19283945/39861745-d381da8c-5441-11e8-8975-727835851c74.png)


Next, I tried running sp_BlitzFirst while running the concurrency scripts we have been using for our tests. I ran `EXEC sp_BlitzFirst @ExpertMode = 1, @Seconds = 30` on sql server while launching teh ParallelTest.js as many times as I could (basically, I launch that script again right after it finishes).
The following image is the results from that:
![image](https://user-images.githubusercontent.com/19283945/39861619-70d513ae-5441-11e8-8b9c-22133a62ab72.png)
Here we have that row with wait_type == LCK_M_S that was absent when running the queries from SQLQueryStress, with more than 2 seconds waited each time. 
We still do not have the RESOURCE_SEMAPHORE waits, so I would guess that we are seeing something different form what's in the post that proposed the testing procedure.

**Comment by MatteoPiovanelli** [comment]
the second result for querying 
nhibernate LCK_M_S
on google is #4566 , so I am checking that

**Comment by MatteoPiovanelli** [comment]
I am trying to see the information from sp_BlitzLock (one of the scripts from the sources @sebastienros posted).

![image](https://user-images.githubusercontent.com/19283945/39871408-00ed2206-5465-11e8-9d0c-849c558109bc.png)

Note that all the deadlock reported there are the ones I caused earlier by launching the scripts for 2 concurrent page creations (see two comments above for my test conditions). 
I don't know why we would have a bunch of deadlocks related to TitlePartRecord and some to ContentItemVersionRecord. Looking at the other results for this stored procedure, it looks like the locks were on TitlePartRecord for the first 9 deadlocks, on ContentItemVersionRecord for the other 13 deadlocks afterwards.

The answer (https://stackoverflow.com/a/5353251/2669614) tells that automatic primary keys are often troublesome. Since in the results here we see them as "involved" in the deadlocks, they may be on to something.

**Comment by sebastienros** [comment]
Can you try to configure NH to use Hilo identity ? This is what I am also doing in YesSql (Orchard Core)

**Comment by MatteoPiovanelli** [comment]
I am checking that right now.

**Comment by MatteoPiovanelli** [comment]
@sebastienros could you point me to wherever the NHibernate configuration is handled right now, so I don't have to search for it myself?

**Comment by MatteoPiovanelli** [comment]
I found that, but I still have to figure out how to configure that setting.
So far, I found this: https://stackoverflow.com/a/11402990/2669614

**Comment by MatteoPiovanelli** [comment]
I don't think I know enough about NHibernate to figure out how to do the configuration for this in a timely manner. Perhaps someone could step in?
I would be adding the stuff from that SO answer in `AbstractDataServicesProvider`, or in `SqlServerDataServicesProvider`, where I would override `AlterConfiguration`.

I don't know what else I would need. 
Likely something to update all existing tables in the db, so that the current Ids are not primary key anymore, but rather "become" computed from the hilo table. This should happen in such a way that the value in the hilo table is already not 1, but rather a value preventig the newly created ids to clash with pre-existing ones.

Also, changes to how new tables are created, so that wherever in a migration we define a column as `PrimaryKey()`, it does what's necessary for the hilo table and script to do their thing.

**Comment by MatteoPiovanelli** [comment]
Thinking about my last comment from yesterday, i think that the changes to the configuration shouldn't be in `SqlServerDataServicesProvider`, because they imply other changes in the behaviour of everything, and I am not sure there is an easy way to make sure it's possible for those to be in place for MSSQL Server only.

**Comment by MatteoPiovanelli** [comment]
Looking at the code, I have more comments on this. What I say here may be wrong, so correct me as needed:

 - `Orchard.ContentManagement.Records.ContentPartAlteration` sets the mappings for the Id of `ContentPartRecord` and `ContentPartVersionRecord` to be generated as a "copy" of the id of the `ContentItemRecord` or `ContentItemVersionRecord` respectively. This may mean that we only need the hilo generation of the Id for those two, and for any record that is not a `ContentPartRecord` nor a `ContentPartVersionRecord`. What I mean is that we would need the hilo generation for all those columns in the db that we currently define somewhat like this: `.Column<int>("Id", column => column.PrimaryKey().Identity())`. I think the change should be in whatever handles the booleans set from `.Identity()` (is it `Orchard.Data.Migration.Interpreters.DefaultDataMigrationInterpreter`?).

 - I am not sure whether we should be thinking about setting this up only for MSSQL Server. While this is an issue we found there, I would think that its solution would not create any issue on other dbms. It seems to me that having this solution coexist with what's currently in live code would create a lot more complexity than "simply" using hilo for every system. This of course assumes that other dbms have no issue with that approach, and I don't see why they would.

**Comment by MatteoPiovanelli** [comment]
I found a workaround for the issues here.
Since the tools report that the SELECT statements are deadlocking because sql server is too pessimistic and locks the entire tables for TitlePartRecord and ContentItemVersionRecord, I wrote an interceptor that adds "WITH (NOLOCK)" to the SELECT statements for those tables. This appears to fix things for the specific case of the issue here, and lets me create as many concurrent pages as I want (Page is defined to have CommonPart, TitlePart, MenuPart).

I'll do a PR here, to share this code, but I don't think it should be merged yet, because I am not sure this approach is correct and does not break something else. Also, this code is WIP: at the very least, if we agree it works, I would add an interface for providers to add further table names.
